File: src/main/java/dev/psygamer/wireframe/event/EventBus.java
package dev.psygamer.wireframe.event;

import dev.psygamer.wireframe.event.api.Event;
import dev.psygamer.wireframe.event.api.EventSubscriber;
import dev.psygamer.wireframe.event.api.IEventBus;
import io.netty.util.internal.ConcurrentSet;
import net.jodah.typetools.TypeResolver;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

public class EventBus implements IEventBus {
	
	private final Class<?> eventMarkerClass;
	
	private final ConcurrentSet<Object> registeredListeners = new ConcurrentSet<>();
	private final ConcurrentHashMap<Class<? extends Event>, List<IEventListener>> listeners = new ConcurrentHashMap<>();
	
	public EventBus() {
		this(Event.class);
	}
	
	public EventBus(final Class<?> eventMarkerClass) {
		this.eventMarkerClass = eventMarkerClass;
	}
	
	@Override
	public void register(final Object target) {
		if (target == null || this.registeredListeners.contains(target))
			return;
		
		this.registeredListeners.add(target);
		
		if (target.getClass() == Class.class) {
			registerClass((Class<?>) target);
		} else {
			registerObject(target);
		}
	}
	
	private void registerClass(final Class<?> clazz) {
		Arrays.stream(clazz.getMethods())
				.filter(method -> Modifier.isStatic(method.getModifiers()))
				.filter(method -> method.isAnnotationPresent(EventSubscriber.class))
				.forEach(method -> addListener(null, method, method.getAnnotation(EventSubscriber.class).priority()));
	}
	
	private void registerObject(final Object object) {
		Arrays.stream(object.getClass().getMethods())
				.filter(method -> !Modifier.isStatic(method.getModifiers()))
				.filter(method -> method.isAnnotationPresent(EventSubscriber.class))
				.forEach(method -> addListener(object, method, method.getAnnotation(EventSubscriber.class).priority()));
	}
	
	@Override
	public <T extends Event> void addListener(final Consumer<T> consumer) {
		addListener(consumer, Event.Priority.NORMAL);
	}
	
	@Override
	public <T extends Event> void addListener(final Consumer<T> consumer, final Event.Priority priority) {
		final Class<T> eventClass = getEventClass(consumer);
		
		if (!eventClass.isAssignableFrom(this.eventMarkerClass))
			throw new IllegalArgumentException("Event must inherit: " + this.eventMarkerClass);
		
		addListener(eventClass,
				new ConsumerEventListener<>(consumer, priority, this)
		);
	}
	
	@Override
	public <T extends Event> void addListener(final Object instance, final Method method) {
		addListener(instance, method, Event.Priority.NORMAL);
	}
	
	@Override
	public <T extends Event> void addListener(final Object instance, final Method method, final Event.Priority priority) {
		if (method.getParameterTypes().length != 1)
			throw new IllegalArgumentException("Event method must have 1 argument");
		if (!method.getParameterTypes()[0].isAssignableFrom(Event.class))
			throw new IllegalArgumentException("Event method's argument must inherit Event");
		
		final Class<T> eventClass = (Class<T>) method.getParameterTypes()[0];
		
		if (!eventClass.isAssignableFrom(this.eventMarkerClass))
			throw new IllegalArgumentException("Event must inherit: " + this.eventMarkerClass);
		
		addListener(eventClass,
				new MethodEventListener(instance, method, priority, this)
		);
	}
	
	private void addListener(final Class<? extends Event> eventClass, final IEventListener listener) {
		if (eventClass == null || listener == null ||
				(this.listeners.contains(eventClass) && this.listeners.get(eventClass).contains(listener))
		) return;
		
		if (!this.listeners.contains(eventClass))
			this.listeners.put(eventClass, new ArrayList<>());
		
		this.listeners.get(eventClass).add(listener);
	}
	
	@Override
	public boolean post(final Event event) {
		final List<IEventListener> eventListeners = this.listeners.get(event.getClass());
		
		eventListeners.stream()
				.filter(listener -> listener.getEventBus() == this)
				.sorted(Comparator.comparing(IEventListener::getPriority))
				.forEach(listener -> listener.invoke(event));
		
		return event.isCancelable() & event.isCanceled();
	}
	
	private <T extends Event> Class<T> getEventClass(final Consumer<T> consumer) {
		final Class<T> eventClass = (Class<T>) TypeResolver.resolveRawArgument(Consumer.class, consumer.getClass());
		
		if ((Class<?>) eventClass == TypeResolver.Unknown.class) {
			throw new IllegalStateException("Failed to resolve consumer event type: " + consumer);
		}
		return eventClass;
	}
}

File: src/main/kotlin/dev/psygamer/wireframe/event/EventBus.kt
package dev.psygamer.wireframe.event

import io.netty.util.internal.ConcurrentSet
import net.jodah.typetools.TypeResolver
import java.lang.reflect.*
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import dev.psygamer.wireframe.event.api.*

class EventBus @JvmOverloads constructor(
	private val eventMarkerClass: Class<*> = Event::class.java,
) : IEventBus {

	private val registeredListeners = ConcurrentSet<Any>()
	private val listeners = ConcurrentHashMap<Class<out Event>, MutableList<IEventListener>>()

	override fun register(target: Any) {
		if (registeredListeners.contains(target))
			return

		registeredListeners.add(target)
		registerTarget(target)
	}

	override fun post(event: Event): Boolean {
		val eventListeners: List<IEventListener> = listeners[event.javaClass]!!

		eventListeners
			.filter { it.eventBus == this }
			.sortedWith(Comparator.comparing(IEventListener::priority))
			.forEach { it(event) }

		return event.isCancelable && event.isCanceled
	}

	override fun <T : Event> addListener(consumer: (T) -> Unit, priority: EventPriority) {
		val eventClass = getEventClass(consumer)
		require(eventClass.isAssignableFrom(eventMarkerClass)) { "Event must inherit: $eventMarkerClass!" }

		addListener(eventClass, ConsumerEventListener(consumer, priority, this))
	}

	override fun addListener(instance: Any?, method: Method, priority: EventPriority) {
		require(method.parameterTypes.size == 1) { "Event method must have exactly 1 argument!" }
		require(method.parameterTypes[0].isAssignableFrom(Event::class.java)) { "Event method's argument must inherit Event!" }
		require(!(instance != null && !Modifier.isStatic(method.modifiers))) { "Class event listeners can only register static methods" }
		require(!(instance == null && Modifier.isStatic(method.modifiers))) { "Instance event listeners can only register member methods" }

		@Suppress("UNCHECKED_CAST") // It's checked...
		val eventClass = method.parameterTypes[0] as Class<out Event>
		require(eventClass.isAssignableFrom(eventMarkerClass)) { "Event must inherit: $eventMarkerClass!" }

		addListener(eventClass, MethodEventListener(instance, method, priority, this))
	}

	private fun addListener(eventClass: Class<out Event>, listener: IEventListener) {
		if (listeners.containsKey(eventClass) && listener in listeners[eventClass]!!)
			return

		if (!listeners.containsKey(eventClass))
			listeners[eventClass] = ArrayList()
		listeners[eventClass]!!.add(listener)
	}

	private fun registerTarget(target: Any) {
		getEventMethodsFromTarget(target).forEach {
			addListener(if (target.javaClass == Class::class.java) null else target, it, it.getAnnotation(EventSubscriber::class.java).priority)
		}
	}

	private fun <T : Event> getEventClass(consumer: (T) -> Unit): Class<T> {
		val eventClass = TypeResolver.resolveRawArgument(Consumer::class.java, consumer.javaClass) as Class<*>
		check(eventClass != TypeResolver.Unknown::class.java) { "Failed to resolve consumer event type: $consumer!" }
		require(eventClass.isAssignableFrom(Event::class.java)) { "Event method's argument must inherit Event!" }

		@Suppress("UNCHECKED_CAST") // It's checked...
		return eventClass as Class<T>
	}
}